### ISSUE TITLE
MonacoEditorReactComp caches text and won't update it even if new text is set in userConfig

### ISSUE THREAD

**Issue by azhakhan** [TypeFox/monaco-languageclient#752]
i use monaco-languageclient to render python code editor

i'm using `MonacoEditorReactComp` and configured it as:
- `useConfigureMonacoWorkers` is copied straight from the docs
- in `CreateUserConfig` i configure `LSP` (websocket url), `text` (which is code pulled form the server), `uri` (full path to the file), and `getKeybindingsServiceOverride` 
- `onTextChanged` is passed to `MonacoEditorReactComp` and it handles text changes in parent component
- in `PythonEditor` (parent component), i have 2 key pieces:
  - `code` - the original code pulled from the server
  - `updatedCode` - a updated version of the `code` (updated via `handleTextChanged` based on changes from `MonacoEditorReactComp`) to be used by parent component

here is how it looks in the code:

```TypeScript
import '@codingame/monaco-vscode-python-default-extension';
import { UserConfig } from 'monaco-editor-wrapper';
import type { TextChanges } from '@typefox/monaco-editor-react';
import { MonacoEditorReactComp } from '@typefox/monaco-editor-react';
import { useWorkerFactory } from 'monaco-editor-wrapper/workerFactory';
import getKeybindingsServiceOverride from '@codingame/monaco-vscode-keybindings-service-override';


const useConfigureMonacoWorkers = () => {
  useWorkerFactory({
    ignoreMapping: true,
    workerLoaders: {
      editorWorkerService: () =>
        new Worker(
          new URL('monaco-editor/esm/vs/editor/editor.worker.js', import.meta.url),
          { type: 'module' },
        ),
    },
  });
};

const CreateUserConfig = ({ text, uri }: { text: string; uri: string }): UserConfig => {
  return {
    languageClientConfig: {
      languageId: 'python',
      options: {
        $type: 'WebSocketUrl',
        url: LSP_URL,
        startOptions: {
          onCall: () => {},
          reportStatus: false,
        },
        stopOptions: {
          onCall: () => {},
          reportStatus: false,
        },
      },
    },
    wrapperConfig: {
      serviceConfig: {
        userServices: {
          ...getKeybindingsServiceOverride(),
        },
        debugLogging: false,
      },
      editorAppConfig: {
        $type: 'extended',
        codeResources: {
          main: { text, uri },
        },
      },
    },
  };
};

interface EditorProps {
  value: string;
  uri: string;
  onTextChanged: (textChanges: TextChanges) => void;
}

const Editor: FC<EditorProps> = ({ value, uri, onTextChanged }) => {
  const config = CreateUserConfig({ text: value, uri });

  return (
    <Stack style={{ height: '100%' }} key={value} pt={4}>
      <MonacoEditorReactComp
        key={value}
        userConfig={config}
        style={{ height: '100%' }}
        onTextChanged={onTextChanged}
      />
    </Stack>
  );
};

export const PythonEditor = () => {
  // pull code saved in file from the server
  const { code, refetch, isFetching, isLoading, isRefetching } = useFile({ ... });
  const { cwd } = useGetPage({ ... });
  // handle updates
  const [updatedCode, setUpdatedCode] = useState<string>('');
  // Create a ref to store the updatedCode
  const updatedCodeRef = useRef(updatedCode);
  updatedCodeRef.current = updatedCode;

  // sync updatedCode with server copy when a new version received 
  useEffect(() => {
    if (code) {
      setUpdatedCode(code);
    }
  }, [code]);
  
  ...

  const handleTextChanged = (textChanges: TextChanges) => {
    const text = textChanges.main;
    setUpdatedCode(text);
  };

  useConfigureMonacoWorkers();

  return (
    <Stack h="full" bg="white" spacing="0" w="full">
        {isFetching || isLoading || isRefetching || !cwd ? (
          <Skeleton startColor="gray.100" endColor="gray.400" h="full" />
        ) : (
          <Editor
            key={code}
            value={code}
            uri={`${cwd}/workspace/${appName}/${pageName}/${fileName}`}
            onTextChanged={handleTextChanged}
          />
        )}
      </Tabs>
    </Stack>
  );
};
```

besides @typefox and @codingame packages, im using chakra ui, react query, and jotai.

editor works if only editor is used.

however, if the `code` is updated by some other components while `MonacoEditorReactComp` has its own changes (unsaved or even saved), even if `useFile` pulls a new version of the `code` and passes it to `MonacoEditorReactComp`, `MonacoEditorReactComp` **will not update its text content** (text).

i can confirm that `Editor` receives a new code with `value={code}` and that `config` is contains the latest `code` in `config.wrapperConfig.editorAppConfig.codeResources.main.text`.

i tried to trigger a new render of `Editor` by passing new `code` as new key, but no luck. 

it is as if `MonacoEditorReactComp` is cached and won’t update it even if new `config` received.

how can i resolve this?

thanks!

**Comment by kaisalmen** [comment]
Hi @azhakhan are you aware of this https://github.com/TypeFox/monaco-languageclient/blob/main/packages/examples/src/python/client/reactPython.tsx#L47-L49 This way you can get hold of the wrapper and do whatever you want afterwards. When you want to update editor text you can do this via wrapper. The react component is basically just a shell around the monaco-editor-wrapper.

**Comment by azhakhan** [comment]
hey @kaisalmen, thanks for the response!

i went over the methods in `MonacoEditorLanguageClientWrapper` and the one that seemed most relevant for my case was `updateCodeResources`, so i updated my code as:
```TypeScript
			<MonacoEditorReactComp
				key={value}
				userConfig={config}
				style={{ height: '100%' }}
				onTextChanged={onTextChanged}
				onLoad={(editorWrapper: MonacoEditorLanguageClientWrapper) => {
					editorWrapper.updateCodeResources({
						main: { text: value, uri },
						original: { text: value, uri },
					});
				}}
			/>
```
my assumption here is that i need to reset `main` and `original` to the newly received code value.

however, the issue persists.

i also tried it with async and initAndStart
```TypeScript
			<MonacoEditorReactComp
				key={`${value}_${uri}`} // unique key based on both value and uri
				userConfig={config}
				style={{ height: '100%' }}
				onTextChanged={onTextChanged}
				onLoad={async (editorWrapper: MonacoEditorLanguageClientWrapper) => {
					try {
						await editorWrapper.initAndStart(
							config,
							document.getElementById('editorContainer'),
						);
						editorWrapper.updateCodeResources({
							main: { text: value, uri },
							original: { text: value, uri },
						});
					} catch (error) {
						console.error('Failed to initialize editor wrapper:', error);
					}
				}}
			/>
```

but still getting caching issue.

am i missing something?

**Comment by kaisalmen** [comment]
@azhakhan The idea is that you store the wrapper that is passed in `onLoad` in your app. You don't have to call `initAndStart`. If you want to update the text (from your appt) you just need to call `updateCodeResources`. `onLoad` let's you know the content of the editor was changed (you typed something).

The latest version of the react-component (not yet released, I will publish a pre-release today or tomorrow and let you know) will perform a full re-init if an updated config is passed.

I hope this helps.

**Comment by azhakhan** [comment]
hey @kaisalmen, thanks for the pointer!

i’ve made the following changes:

- created `wrapperRef` to store reference to `MonacoEditorLanguageClientWrapper`
- assign `wrapperRef.current = editorWrapper` when onLoad is called
- update `wrapperRef.current.updateCodeResources` when value changes using useEffect
- added `isEditorLoaded` flag to ensure `wrapperRef.current` updated only after editor is loaded

```tsx
const Editor: FC<EditorProps> = ({ value, uri, onTextChanged }) => {
	const wrapperRef = useRef<MonacoEditorLanguageClientWrapper | null>(null);
	const [isEditorLoaded, setIsEditorLoaded] = useState(false);
	const [config, setConfig] = useState(() => CreateUserConfig({ text: value, uri }));

	useEffect(() => {
		setConfig(CreateUserConfig({ text: value, uri }));
	}, [value, uri]);

	useEffect(() => {
		if (isEditorLoaded && wrapperRef.current) {
			console.log('WRAPPERREF.CURRENT', wrapperRef.current);
			wrapperRef.current.updateCodeResources({
				main: { text: value, uri },
				original: { text: value, uri },
			});
		}
	}, [isEditorLoaded, value, uri]);

	const handleLoad = useCallback(async (editorWrapper: MonacoEditorLanguageClientWrapper) => {
		console.log('EDITORWRAPPER', editorWrapper);
		wrapperRef.current = editorWrapper;
		setIsEditorLoaded(true);
	}, []);

	return (
		<Stack style={{ height: '100%' }} key={`${uri}-editor`} pt={4}>
			<MonacoEditorReactComp
				key={`${uri}-editor`}
				userConfig={config}
				style={{ height: '100%' }}
				onTextChanged={onTextChanged}
				onLoad={handleLoad}
			/>
		</Stack>
	);
};
```

i can confirm that `editorWrapper` and `wrapperRef.current` have the same values

![image](https://github.com/user-attachments/assets/e913e739-d140-4731-bc60-46990c2b4f3e)

i also confirmed that `wrapperRef.current` has the right data in `editorApp.config.codeResources.main.text`

![image](https://github.com/user-attachments/assets/c9bac2c8-8725-4556-b7d2-81870b936556)

and the right `MonacoEditorLanguageClientWrapper` (id 10 in this case) is mounted to the Editor

![image](https://github.com/user-attachments/assets/1f1bee3b-b4a8-4ff9-88e2-e14dff7955ae)

however, the old value is still shown in the monaco editor (see options in Select, only shown instead of 3)

![image](https://github.com/user-attachments/assets/8d08ddba-1a70-4dff-b177-a849144cd193)
 
not sure what else to try here.

i'll wait for the latest version of the react-component and try again.

thanks again for helping me with this!

**Comment by kaisalmen** [comment]
@azhakhan https://www.npmjs.com/package/@typefox/monaco-editor-react/v/6.0.0-next.1 is now available. This is a pre-release and introduces some changes to the wrapper configuration.

For reference take a look at the updated python example:
https://github.com/TypeFox/monaco-languageclient/blob/main/packages/examples/src/python/client/reactPython.tsx
https://github.com/TypeFox/monaco-languageclient/blob/main/packages/examples/src/python/client/config.ts

**Comment by kaisalmen** [comment]
@azhakhan can you share a reproducible example in a repo or can you augment a react example in this repo so your problem can be reproduced? Thank you.

**Comment by Koxx3** [comment]
hello, any solution on this ?

**Comment by kaisalmen** [comment]
@azhakhan v6 is available. If this issue reoccurs, please open a new issue.
